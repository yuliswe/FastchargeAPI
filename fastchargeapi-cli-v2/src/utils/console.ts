import { createInterface } from "readline";

export const print = (message: string) => {
  process.stdout.write(message);
};

export const println = (message: string) => {
  print(message + "\n");
};

export const readline = (prompt: string) =>
  new Promise<string>((resolve) => {
    const rl = createInterface({
      input: process.stdin,
      output: process.stdout,
    });

    rl.question(prompt, (answer) => {
      resolve(answer);
      rl.close();
    });
  });
