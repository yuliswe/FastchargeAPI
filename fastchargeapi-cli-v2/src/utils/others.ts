export const parseEnum =
  <T extends string>(enumObject: Record<string, T>) =>
  (value: string): T => {
    if (!Object.values(enumObject).includes(value as T)) {
      throw new Error(`Invalid value. Must be one of ${Object.values(enumObject).join(", ")}`);
    }
    return value as T;
  };
