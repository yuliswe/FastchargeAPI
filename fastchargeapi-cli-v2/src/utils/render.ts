import { Decimal } from "decimal.js";

export function renderMoney(amount: string | number | Decimal) {
  return Number(amount).toLocaleString("en-US", {
    style: "currency",
    currency: "USD",
  });
}
