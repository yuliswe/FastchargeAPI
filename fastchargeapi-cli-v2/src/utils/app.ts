import { graphql } from "src/__generated__/gql";
import { GQLClient } from "src/graphqlClient";
import { NotFoundSimpleGQLError } from "src/simplifiedGQLErrors";

export async function getAppPKByPKOrName(args: { client: GQLClient; pkOrName: string }) {
  const { client, pkOrName } = args;
  try {
    const result = await client.mutate({
      mutation: graphql(`
        mutation GetAppPKByName($appName: String!) {
          getAppByName(name: $appName) {
            pk
          }
        }
      `),
      variables: {
        appName: pkOrName,
      },
    });
    return result.data.getAppByName.pk;
  } catch (e) {
    if (e instanceof NotFoundSimpleGQLError) {
      const result = await client.mutate({
        mutation: graphql(`
          mutation GetAppPKById($pk: ID!) {
            getApp(pk: $pk) {
              pk
            }
          }
        `),
        variables: {
          pk: pkOrName,
        },
      });
      return result.data.getApp.pk;
    }
    throw e;
  }
}
