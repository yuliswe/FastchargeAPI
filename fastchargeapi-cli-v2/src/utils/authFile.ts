import { promises as fs } from "fs";
import { glob } from "glob";
import { decodeProtectedHeader, importX509, errors as joseErrors, jwtVerify } from "jose";
import fetch from "node-fetch";
import os from "os";
import path from "path";
import { graphql } from "src/__generated__/gql";
import { authServiceHost } from "src/env";
import { getGQLClient } from "src/graphqlClient";
import { tiChecker } from "src/tiChecker";
import { AuthFileContent, RefreshIdTokenResult, VerifyIdTokenResult } from "src/types/authFile";

export async function getGoogleCert(): Promise<{ [hash: string]: string }> {
  const response = await fetch(
    "https://www.googleapis.com/robot/v1/metadata/x509/securetoken@system.gserviceaccount.com"
  );
  return (await response.json()) as { [hash: string]: string };
}

export function getAuthFilePath(profile?: string): string {
  const home = os.homedir();
  if (profile) {
    return path.join(home, ".fastcharge", `auth.${profile}.json`);
  }
  return path.join(home, ".fastcharge", "auth.json");
}

export async function listAuthFiles(): Promise<string[]> {
  const home = os.homedir();
  return await glob([path.join(home, ".fastcharge", "auth.*.json"), path.join(home, ".fastcharge", "auth.json")]);
}

export async function deleteAuthFile(profile?: string) {
  const authFilePath = getAuthFilePath(profile);
  await fs.access(authFilePath);
  await fs.unlink(authFilePath);
}

export async function readAuthFile(profile?: string): Promise<AuthFileContent> {
  const authFilePath = getAuthFilePath(profile);
  const authFileContent = await fs.readFile(authFilePath, "utf-8");
  const auth = tiChecker.AuthFileContent.from(JSON.parse(authFileContent));
  auth.issuer = auth.issuer || "firebase";
  return auth;
}

/**
 * Partially update the auth file context, or create a new auth file with the
 * specified content.
 */
export async function writeToAuthFile(
  profile: string | undefined,
  content: Partial<AuthFileContent>
): Promise<AuthFileContent> {
  const authFilePath = getAuthFilePath(profile);
  await fs.mkdir(path.dirname(authFilePath), { recursive: true });
  let currentAuthFile;
  try {
    currentAuthFile = await readAuthFile(profile);
  } catch {
    currentAuthFile = null;
  }
  // default issuer to firebase
  const newAuthFile = { issuer: "firebase", ...currentAuthFile, ...content };
  await fs.writeFile(authFilePath, JSON.stringify(newAuthFile, null, 2));
  await fs.chmod(authFilePath, 0o600);
  return tiChecker.AuthFileContent.from(newAuthFile);
}

export type VerifyOrRefreshIdTokenResult = {
  refreshed: boolean;
  idToken: string;
  refreshToken: string;
  email: string;
};

export async function refreshIdToken(refreshToken: string): Promise<RefreshIdTokenResult> {
  const resp = await fetch(`${authServiceHost}/refresh-idtoken`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ refreshToken: refreshToken }),
  });

  if (!resp.ok) {
    throw new Error("Unable to refresh id token. " + (await resp.text()));
  }

  return tiChecker.RefreshIdTokenResult.from(await resp.json());
}

export async function verifyIdToken(idToken: string): Promise<VerifyIdTokenResult> {
  const { kid, alg } = decodeProtectedHeader(idToken);

  if (!kid) {
    throw new Error("No kid found in id token.");
  }

  if (!alg) {
    throw new Error("No alg found in id token.");
  }

  const certs = await getGoogleCert();
  const cert = certs[kid];
  if (!cert) {
    throw new Error("No cert found for kid. Google may have rotated the cert and your id token is expired.");
  }

  const key = await importX509(cert, alg);

  try {
    const { payload } = await jwtVerify(idToken, key);
    return tiChecker.VerifyIdTokenResult.from(payload);
  } catch (error) {
    if (error instanceof joseErrors.JWTExpired) {
      throw new Error("Id token expired.");
    }
    throw new Error("Invalid id token. ");
  }
}

export async function verifyOrRefreshIdToken(args: {
  idToken: string;
  refreshToken: string;
  forceRefresh?: boolean;
}): Promise<VerifyOrRefreshIdTokenResult> {
  const { idToken, refreshToken, forceRefresh } = args;
  if (forceRefresh) {
    const newToken = await refreshIdToken(refreshToken);
    const { email } = await verifyIdToken(newToken.idToken);
    return {
      email,
      idToken: newToken.idToken,
      refreshToken: newToken.refreshToken,
      refreshed: true,
    };
  } else {
    try {
      const { email } = await verifyIdToken(idToken);
      return {
        email,
        idToken,
        refreshToken,
        refreshed: false,
      };
    } catch (e) {
      const newToken = await refreshIdToken(refreshToken);
      const { email } = await verifyIdToken(newToken.idToken);
      return {
        email: email,
        idToken: newToken.idToken,
        refreshToken: newToken.refreshToken,
        refreshed: true,
      };
    }
  }
}

/**
 * Gets the id token from the auth file. If the file doesn't exist, returns
 * None. If the id token is issued by Firebase, verifies the token. If the id
 * token is invalid, refreshes it and returns the new id token. If unable to
 * refresh, return None. If the id token is not issued by FastchargeAPI, skips
 * the verification. The new id token and refresh token are written to the auth
 * file.
 */
export async function readOrRefreshAuthFile(args: {
  profile?: string;
  forceRefresh?: boolean;
}): Promise<AuthFileContent | null> {
  const { profile, forceRefresh } = args;
  const auth = await readAuthFile(profile);
  if (auth) {
    if (auth.issuer === "firebase") {
      const user = await verifyOrRefreshIdToken({ ...auth, forceRefresh });
      if (user && user.refreshed) {
        const resp = await getGQLClient(user).query({
          query: graphql(`
            query GetUserPKByEmail($email: Email!) {
              getUserByEmail(email: $email) {
                pk
              }
            }
          `),
          variables: { email: user.email },
        });
        const userPK = resp.data.getUserByEmail.pk;
        return await writeToAuthFile(profile, {
          idToken: user.idToken,
          refreshToken: user.refreshToken,
          email: user.email,
          userPK,
        });
      } else {
        return auth;
      }
    } else if (auth.issuer === "fastchargeapi") {
      return auth;
    } else {
      throw new Error(`Unknown issuer: ${auth.issuer}`);
    }
  }
  return null;
}
