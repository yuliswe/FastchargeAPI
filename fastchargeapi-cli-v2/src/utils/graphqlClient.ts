import { getGQLClient } from "src/graphqlClient";
import { readOrRefreshAuthFileContentOrExit } from "src/utils/authFile";

export async function getAuthorizedGQLClient(args: { profile?: string }) {
  const { profile } = args;
  const authFile = await readOrRefreshAuthFileContentOrExit({ profile });
  return getGQLClient(authFile);
}
