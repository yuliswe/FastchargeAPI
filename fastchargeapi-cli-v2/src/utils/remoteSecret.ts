import crypto from "crypto";
import { EncryptJWT, JWTPayload, SignJWT, jwtDecrypt, jwtVerify } from "jose";
import { graphql } from "src/__generated__/gql";
import { getGQLClient } from "src/graphqlClient";

export async function getRemoteSecret(args: {
  key: string;
  jweSecret: Uint8Array;
  jwtSecret: Uint8Array;
}): Promise<object> {
  const { key, jweSecret, jwtSecret } = args;
  const resp = await getGQLClient().query({
    query: graphql(`
      query GetSecret($key: String!) {
        getSecret(key: $key) {
          value
        }
      }
    `),
    variables: { key },
  });

  const secret = resp.data.getSecret.value;

  const jwtVerified = await jwtVerify(secret, jwtSecret, { algorithms: ["HS512"] });

  const { encrypted } = jwtVerified.payload as { encrypted: string };
  const jweDecrypted = await jwtDecrypt(encrypted, jweSecret);
  return jweDecrypted.payload;
}

export async function setRemoteSecret(args: {
  key: string;
  value: JWTPayload;
  description?: string;
  expireAt?: number;
  jweSecret: Uint8Array;
  jwtSecret: Uint8Array;
}) {
  const { key, value, description, expireAt, jweSecret, jwtSecret } = args;
  const signedValue = await encryptAndSign(value, { jweSecret, jwtSecret });
  const response = await getGQLClient().mutate({
    mutation: graphql(`
      mutation PutSecret($key: String!, $signedValue: String!, $description: String, $expireAt: Timestamp) {
        createSecret(key: $key, value: $signedValue, description: $description, expireAt: $expireAt) {
          createdAt
        }
      }
    `),
    variables: {
      key,
      signedValue,
      description,
      expireAt,
    },
  });
  return response;
}

export async function encryptAndSign(
  body: JWTPayload,
  options: {
    jweSecret: Uint8Array;
    jwtSecret: Uint8Array;
  }
): Promise<string> {
  const { jweSecret, jwtSecret } = options;
  const encrypted = await new EncryptJWT(body)
    .setIssuedAt()
    .setIssuer("fastchargeapi.com")
    .setAudience("fastchargeapi.com")
    .setProtectedHeader({
      alg: "dir",
      enc: "A256CBC-HS512",
    })
    .encrypt(jweSecret);

  const signed = await new SignJWT({
    encrypted,
  })
    .setProtectedHeader({
      alg: "HS512",
    })
    .sign(jwtSecret);
  return signed;
}

export function createSecret(): Uint8Array {
  const secret = new Uint8Array(64);
  crypto.getRandomValues(secret);
  return secret;
}

export function createSecretPair(): { jweSecret: Uint8Array; jwtSecret: Uint8Array } {
  return { jweSecret: createSecret(), jwtSecret: createSecret() };
}
