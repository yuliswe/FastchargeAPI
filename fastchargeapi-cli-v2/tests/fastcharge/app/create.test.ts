import { AppTableIndex } from "@/src/database/models/App";
import { User } from "@/src/database/models/User";
import { UserPK } from "@/src/pks/UserPK";
import { createTestApp } from "@/tests/test-data/App";
import { createTestUser } from "@/tests/test-data/User";
import { baseRequestContext as context } from "@/tests/test-utils/test-utils";
import { fastcharge, mockLoggedInAsUser } from "tests/utils";
import * as uuid from "uuid";

describe("fastcharge app create --help", () => {
  it("prints help message", async () => {
    const { stdout, exitCode } = await fastcharge(["app", "create", "--help"]);
    expect(exitCode).toBe(0);
    expect(stdout.getOutput()).toMatchSnapshot();
  });
});

describe("fastcharge app create", () => {
  let testUser: User;

  beforeEach(async () => {
    testUser = await createTestUser(context);
    await mockLoggedInAsUser(testUser);
  });

  it("creates app", async () => {
    const userApp = await context.batched.App.getOrNull(
      {
        owner: UserPK.stringify(testUser),
      },
      {
        using: AppTableIndex.Owner,
      }
    );
    expect(userApp).toBeNull();
    const appName = `testapp-${uuid.v4()}`;
    const { stdout, exitCode } = await fastcharge(["app", "create", "--name", appName]);
    expect(
      stdout.getOutput({
        redactWord: {
          appName: (word) => word.startsWith("testapp-"),
          url: (word) => word.startsWith("http"),
        },
      })
    ).toMatchSnapshot();
    expect(exitCode).toBe(0);
    context.batched.App.clearCache();
    const createdUserApp = await context.batched.App.many(
      {
        owner: UserPK.stringify(testUser),
      },
      {
        using: AppTableIndex.Owner,
      }
    );
    expect(createdUserApp).toMatchSnapshotExceptForProps([
      {
        owner: UserPK.stringify(testUser),
        name: appName,
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      },
    ]);
  });

  it("fails when app already exists", async () => {
    const { name } = await createTestApp(context, {
      owner: UserPK.stringify(testUser),
    });
    const { stdout, exitCode } = await fastcharge(["app", "create", "--name", name]);
    expect(stdout.getOutput()).toMatchSnapshot();
    expect(exitCode).toBe(1);
  });

  it("fails when creating app with invalid app name", async () => {
    const { stdout, exitCode } = await fastcharge(["app", "create", "--name", "&!#!"]);
    expect(exitCode).toBe(1);
    expect(stdout.getOutput()).toMatchSnapshot();
  });

  it("fails when creating app with long app name", async () => {
    const { stdout, exitCode } = await fastcharge(["app", "create", "--name", "a".repeat(70)]);
    expect(exitCode).toBe(1);
    expect(stdout.getOutput()).toMatchSnapshot();
  });

  it("fails when creating app with reserved app name", async () => {
    const { stdout, exitCode } = await fastcharge(["app", "create", "--name", "api"]);
    expect(exitCode).toBe(1);
    expect(stdout.getOutput()).toMatchSnapshot();
  });
});
