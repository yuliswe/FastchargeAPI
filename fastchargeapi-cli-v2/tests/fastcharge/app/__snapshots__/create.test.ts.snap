// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fastcharge app create --help prints help message 1`] = `
"Usage: fastcharge app create [options]

Create a new app

Options:
  --name <name>    Name of the app
  --title <title>  Title of the app, defaults to name
  --help           Display help for command
"
`;

exports[`fastcharge app create creates app 1`] = `
"App created: <REDACTED:appName>
[34m <REDACTED:url>
"
`;

exports[`fastcharge app create creates app 2`] = `
[
  {
    "createdAt": Any<Date>,
    "deleted": false,
    "gatewayMode": "proxy",
    "name": Any<String>,
    "owner": Any<String>,
    "updatedAt": Any<Date>,
    "visibility": "public",
  },
]
`;

exports[`fastcharge app create fails when app already exists 1`] = `
"An app with this name already exists: <REDACTED:appName>
Please use a different name.
"
`;

exports[`fastcharge app create fails when creating app with invalid app name 1`] = `
"Invalid app name "&!#!": can only contain lowercase letters, numbers, and dashes. ([a-z\\d][a-z\\d\\-]*[a-z\\d]).
[33mPlease use a different name.[39m
"
`;

exports[`fastcharge app create fails when creating app with long app name 1`] = `
"Invalid app name "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa": cannot be more than 63 characters.
[33mPlease use a different name.[39m
"
`;

exports[`fastcharge app create fails when creating app with reserved app name 1`] = `
"Invalid app name "api": is a reserved name.
[33mPlease use a different name.[39m
"
`;
